engines/mm/shared/xeen/sound_driver_mt32.cpp

```diff
@@ -0,0 +1,477 @@
+	assert((note & 0x1F) < sizeof(MIDI_NOTE_MAP));
+	uint8 freq = (note & 0xE0) >> 2;
+	uint8 result = MIDI_NOTE_MAP[note & 0x1F];
+	result+= freq + (freq >> 1);
+	return result;
+}
+
+void SoundDriverMT32::pausePostProcess() {
+	if (_field180 && ((_field181 += _field180) < 0)) {
+		if (--_field182 < 0) {
+			_streams[stMUSIC]._playing = false;
+			_field180 = 0;
+			resetFrequencies();
+		}
+	}
+}
+
+void SoundDriverMT32::resetFX() {
+	// notes off on 8ch
+	write(0xB8, 0x7B, 0x00);
+	// pitch bend on 8ch
+	write(0xE8, 0x00, 0x40);
+}
+
+void SoundDriverMT32::resetFrequencies() {
+	// set pitch
+	for (int idx = CHANNEL_COUNT - 1; idx >= 0; idx--) {
+		write(0xE1 + idx, 0x00, 0x40);
+	}
+	// set pan
+	for (int idx = CHANNEL_COUNT - 1; idx >= 0; idx--) {
+		write(0xB1 + idx, 0x0A, 0x3F);
+	}
+	// notes off
+	for (int idx = CHANNEL_COUNT - 1; idx >= 0; idx--) {
+		write(0xB1 + idx, 0x7B, 0x00);
+	}
+}
+
+void SoundDriverMT32::playInstrument(byte channelNum, const byte *data, bool isFx) {
+	debugC(2, kDebugSound, "---START-playInstrument - %d", channelNum);
+
+	write(0xC1 + channelNum, *data, 0x00);
+
+	debugC(2, kDebugSound, "---END-playInstrument");
+}
+
+bool SoundDriverMT32::musSetInstrument(const byte *&srcP, byte param) {
+	srcP += 24;
+	debugC(3, kDebugSound, "musSetInstrument %d -> %x", param, *srcP);
+	_musInstrumentPtrs[param] = srcP;
+	srcP += 2;
+
+	return false;
+}
+
+bool SoundDriverMT32::musSetPitchWheel(const byte *&srcP, byte param) {
+	byte pitch1 = *srcP++;
+	byte pitch2 = *srcP++;
+	debugC(3, kDebugSound, "musSetPitchWheel");
+	write(0xE1 + param, pitch1, pitch2);
+
+	return false;
+}
+
+bool SoundDriverMT32::musSetPanning(const byte *&srcP, byte param) {
+	byte pan = *srcP++;
+	debugC(3, kDebugSound, "mmusSetPanning");
+	write(0xB1 + param, 0x0A, pan);
+
+	return false;
+}
+
+bool SoundDriverMT32::musFade(const byte *&srcP, byte param) {
+	byte note = *srcP++;
+	debugC(3, kDebugSound, "musFade: %x", note);
+	if (param != 8)
+		note = noteMap(note);
+
+	write(0x81 + param, note & 0x7F, 0x40);
+	last_notes[param] = 0xFF;
+
+	return false;
+}
+
+bool SoundDriverMT32::musStartNote(const byte *&srcP, byte param) {
+	byte note = *srcP++;
+	byte fade = *srcP++;
+
+	debugC(3, kDebugSound, "musStartNote %x, %x", note, fade);
+	if (param != 8) {
+		note = noteMap(note);
+	}
+
+	if (param != 8) {
+		if (param != 7)
+			write(0x81 + param, last_notes[param] & 0x7F, 0x7F);
+		else
+			write(0x81 + param, note & 0x7F, 0x7F);
+	}
+
+#if defined(MT32_ADJUST_VOLUME)
+	byte level = calculateLevel(fade, false);
+	if (level > 0)
+		write(0x91 + param, note & 0x7F, level);
+#else
+	write(0x91 + param, note & 0x7F, fade);
+#endif
+	last_notes[param] = note & 0x7F;
+
+	return false;
+}
+
+bool SoundDriverMT32::musSetVolume(const byte *&srcP, byte param) {
+	debugC(3, kDebugSound, "musSetVolume %d", (int)*srcP);
+
+	uint8 status = *srcP++;
+	uint8 volume = *srcP++;
+
+	if (status == 0 && !_field180) {
+		// TODO: this is used in fxEndSubroutine in rolmus
```

Comment:
I don't understand this TODO (and the others below that mention the same). Could you please add some more info, and explain what needs to be done here?

Reply:
It was mentioned in the commit message. In original rolmus driver in fxEndSubroutine there is a comparision with these saved values and if they aren't less than 40 then they gradually decremented issuing additional midi expression commands until they are less than 40. I didn't find where this logic makes a distinct difference so it is not currently implemented and these are there just as reminder of missing non-essential logic.

Reply:
Added previously missing postprocessing functionality so these TODOs are resolved now.